### PR TITLE
Fix BPM <-> MIDI tempo conversions

### PR DESCRIPTION
**Pull request by superbock**
This PR addresses #102 and contains the fixes needed to correctly consider the time signature when converting between BPM (beats per minute) and MIDI tempo (microseconds per quarter note).

Before being able to be merged, these points should be discussed/addressed:

- [ ] `ticks_per_beat` is a bit misleading since it rather should be `ticks_per_quarter_note`. Renaming the variable, however, may break existing code, so this is a rather bad option. Maybe adding a new variable (and deprecate the old one, but keep it around for backwards compatibility) is a better approach.
- [ ] Tests for Python 2 fail at the moment since Python 2 and 3 round differently. This was revealed by adding tests for corner cases.
- [ ] Extend the documentation to reflect all introduced changes, especially update the image in `midi_files.rst`.
- [ ]  Add more descriptive docstrings, but which docstring format should be used?

### CODE REVIEW

File: mido/midifiles/units.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def second2tick(second, ticks_per_beat, tempo):
     note) and tempo (microseconds per beat).
     """
     scale = tempo * 1e-6 / ticks_per_beat
-    return second / scale
+    return int(round(second / scale))
```
Comment (carlthome) [inline mido/midifiles/units.py]: This change will break a program of mine where I let the user decrease the PPQ of MIDI files.

```py
ticks = int(round(mido.second2tick(message.time, ppq, tempo)))
assert ticks > 0 or message.time == 0, "New PPQ is too low."
```

Comment (superbock) [reply mido/midifiles/units.py]: It would be good to see which part of the assert fails (I am not sure if this makes any difference, though), then it's easier to discuss this further.

The rationale behind this change is that "the MIDI standard" requires QQN to be an integer. And until now this function could return a float value.

I am not sure why ticks should be greater 0 or message timings should not be 0. Both are legitimate values as far as I can see. What's the intention behind these checks?

The problem seems to be that different Python versions handle the /-operator differently. Anyways, I see that this change breaks code (although I never thought it would), but IMHO the question remains if your assertion assumes the "right thing".

Comment (carlthome) [reply mido/midifiles/units.py]: > The rationale behind this change is that "the MIDI standard" requires QQN to be an integer. And until now this function could return a float value.

How should we handle the case when second will be truncated to zero for the given PPQ and tempo (or heavily shifted in time, `ceil` doesn't really help)?

> What's the intention behind these checks?

The assertion checks that there will be at least one tick with the given `ppq` when `message.time` is larger than 0.0 seconds. 

When PPQ is chosen too low there could be messages that don't move the tick count at all even though they should, which causes a slight global drift AFAIK.

The problem stems from MAPS having ridiculously high PPQ (https://github.com/craffel/pretty-midi/issues/112), while just setting it to 960 is fine because no message actually seems to require that much time resolution to fit into a piano roll (although the messages might shift just slightly forward and/or backwards in time, which I don't bother to check).

Comment (carlthome) [reply mido/midifiles/units.py]: Btw, what does QQN stand for?

Comment (superbock) [reply mido/midifiles/units.py]: Typo, TQN = ticks per quarter note.

Comment (superbock) [reply mido/midifiles/units.py]: High PPQ (e.g. in the MAPS dataset) should be fine, since they result in large tick values. On the other hand, having low PPQ values _is_ a (known) problem, but I am still not sure how this is affected by the proposed change.

> The assertion checks that there will be at least one tick with the given ppq when message.time is larger than 0.0 seconds.

Ok, but I don't see how moving the `int(round(...))` thing inside the function should alter the result since you apply exactly the same rounding and casting to int by yourself. Are you sure that this change is the culprit and not something else?

Comment (carlthome) [reply mido/midifiles/units.py]: > Ok, but I don't see how moving the int(round(...)) thing inside the function should alter the result since you apply exactly the same rounding and casting to int by yourself.

Haha, you're absolutely right... Sorry. :laughing: 

image

Anyway, could we have something like this internally maybe?

```python
if second != 0 and tick == 0:
    warnings.warn(f"{second} seconds is zero ticks at {ppq} PPQ.")
```

**Comment by carlthome** [comment]
Status?

File: mido/midifiles/units.py [hunk 2]
```diff
@@ -12,35 +12,31 @@ def tick2second(tick, ticks_per_beat, tempo):
...
+def tempo2bpm(tempo, time_signature=(4, 4)):
+    """Convert MIDI file tempo (microseconds per quarter note) to BPM (beats
+    per minute).
 
-        250000 => 240
-        500000 => 120
-        1000000 => 60
+    Depending on the chosen time signature a bar contains a different number of
+    beats. These beats are multiples/fractions of a quarter note, thus the
+    returned tempo depends on the time signature.
```
Comment (mgrachten) [inline mido/midifiles/units.py]: I think it is good to mention here that what constitutes a beat is defined by the denominator of the time signature, and give an example: in 2/4, 3/4, 4/4, ... the beat is a quarter note. In 3/8, 6/8, ... the beat is an eight note, etc.

**Review by rdoursenaud — CHANGES_REQUESTED** [review]

**Comment by rdoursenaud** [comment]
I have added the fix I was suggesting in #454 and merged your changes.

Thank you for your contribution!

**Comment by superbock** [comment]
Thanks a lot!
